http/transport-component.md

```diff
@@ -0,0 +1,95 @@
+# HTTP Transport Component <!-- omit in toc -->
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-05-31  |
+
+Authors: [@marcopolo]
+
+Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]
+
+[@marcopolo]: https://github.com/marcopolo
+[@mxinden]: https://github.com/mxinden
+[@marten-seemann]: https://github.com/marten-seemann
+
+## Table of Contents <!-- omit in toc -->
+- [Context](#context)
+- [What is an HTTP transport](#what-is-an-http-transport)
+- [Multiaddr representation](#multiaddr-representation)
+- [HTTP Paths (and other HTTP Semantics)](#http-paths-and-other-http-semantics)
+  - [Recommendation on including HTTP semantics in multiaddrs](#recommendation-on-including-http-semantics-in-multiaddrs)
+
+
+## Context
+
+This document is only about advertising support for an HTTP transport. It
+doesn't make any assertions about how libp2p should interact with that
+transport. That will be defined in a future document.
+
+This exists to clarify the role of the `/http` component in Multiaddrs early to
+avoid confusion and conflicting interpretations.
+
+## What is an HTTP transport
+
+An HTTP transport is simply a node that can speak some standardized version of
+HTTP. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.
+
+Most environments will have a way to create an HTTP Client and Server, and the
+specific HTTP version used will be opaque. We use the `/http` component at the
+end of the multidadr to signal that this server supports an HTTP transport. The
+end user agent decides on HTTP version to use, based on the multiaddr prefix,
+application, server negotiation, and specific use case. This follows what
+existing `http://` URL implementations do.
+
+## Multiaddr representation
+
+The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
+by information that would let an HTTP client know how to reach the server
+(remember that multiaddrs are [interpreted right to
+left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 
+
+The following are examples of multiaddrs for HTTP transport capable nodes:
+
+* `/dns/example.com/tls/http`
```

Comment:
If this says `/tls`, does that mean we expect both ends to conform to the libp2p TLS spec with its mandatory extension for signing the public key? Or is this vanilla TLS like a browser would perform with a server?

If this would not end in `/http`, it would definitely mean the former, correct?

But that would make it impossible for existing HTTP clients to connect as they don't understand our mandatory extension.

The same question applies to `/quic-v1`.

I think technically, we could say that - because multiaddresses are interpreted from right to left - the `/http` component defines that `/tls` and `/quic-v1` mean vanilla TLS without our extension. This sounds very confusing to me though.

Reply:
> If this says /tls, does that mean we expect both ends to conform to the libp2p TLS spec with its mandatory extension for signing the public key? Or is this vanilla TLS like a browser would perform with a server?

The latter, vanilla TLS. Authentication will be optional in the HTTP transport.

> If this would not end in /http, it would definitely mean the former, correct?

I think so because we only support authenticated streams. And we only have defined the libp2p TLS extension as the auth mechanism.

The libp2p TLS extension is about authentication. If we require auth then we need it (unless we define an alternate way to do auth). In the HTTP transport, auth will be optional so we don't require the libp2p TLS extension.

> The same question applies to /quic-v1.

Yup, and hopefully this same answer makes sense.

Reply:
I just realized that the TLS spec only applies to the `/tls/1.0.0` stream protocol and not to the multiaddr transport component.

However, for QUIC, we explicitly define that the TLS authentication is required if the `/quic-v1` component is used: https://github.com/libp2p/specs/tree/master/quic#peer-authentication

I think this confusion would not arise if we would define `/http1`, `/http2` and `/http3`.

Somewhere else you say that we only want to define that a node speaks the HTTP semantics but since we are talking about transports in the multiaddr, wouldn't it make sense to also talk about the various HTTP versions (which are transport related?).

Reply:
> > However, for QUIC, we explicitly define that the TLS authentication is required if the /quic-v1 component is used: [`master`/quic#peer-authentication](https://github.com/libp2p/specs/tree/master/quic?rgh-link-date=2023-06-15T17%3A46%3A48Z#peer-authentication)
> 
> This spec is specific for libp2p streams on QUIC. `/quic-v1` in the context of `/http` can have its own semantics (e.g. optional peer auth).

At least to me, that is completely non-obvious to me from reading the specs.

Multiaddress components currently always mean the same thing, right? This spec would be the first one where we _change_ the meaning of an existing component.

Reply:
Added in e71d7c5